### PR TITLE
CXP-942: Replace ScopeMapper by a registry

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public function getAcls(string $scopeName): array;
+
+    /**
+     * @return array{
+     *      icon: string,
+     *      type: string,
+     *      entities: string,
+     * }
+     */
+    public function getMessage(string $scope): array;
```
Comment (masacc) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php:30]:
```suggestion
    public function getMessage(string $scopeName): array;
```

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * @return string[]
+     */
+    public function getLowerHierarchyScopes(string $scope): array;
```
Comment (masacc) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php:35]:
```suggestion
    public function getLowerHierarchyScopes(string $scopeName): array;
```

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                throw new \InvalidArgumentException(
+                    \sprintf(
+                        '%s needs only %s',
+                        self::class,
+                        ScopeMapperInterface::class
+                    )
+                );
+            }
+            foreach ($scopeMapper->getAllScopes() as $scope) {
+                $scopesToMapper[$scope] = $entity;
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:34]: throw if scope already exist

File: src/Akeneo/Tool/Bundle/ApiBundle/Resources/config/security.yml [hunk 4]
```diff
@@ -23,8 +23,9 @@ services:
             - '%env(APP_SECRET)%'
             - '%env(APP_SECRET)%'
 
-    pim_api.security.scope_mapper:
-        class: Akeneo\Tool\Bundle\ApiBundle\Security\ScopeMapper
+    Akeneo\Tool\Component\Api\Security\ScopeMapperRegistry:
+        arguments:
+            - !tagged_iterator { tag: 'pim_api.security.scope_mapper', index_by: 'entity' }
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Bundle/ApiBundle/Resources/config/security.yml:28]:
```suggestion
            - !tagged_iterator { tag: 'pim_api.security.scope_mapper'}
```

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    public function getAllScopes(): array
+    {
+        $scopes = [];
+        foreach ($this->scopeMappers as $scopeMapper) {
+            \array_push($scopes, ...$scopeMapper->getAllScopes());
+        }
+
+        return \array_values(\array_unique($scopes));
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:48]:
```suggestion
        return \array_keys($this->scopesToMapper);
```

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                        ScopeMapperInterface::class
+                    )
+                );
+            }
+            foreach ($scopeMapper->getAllScopes() as $scope) {
+                $scopesToMapper[$scope] = $entity;
+            }
+        }
+        $this->scopeMappers = $scopeMappers;
+        $this->scopesToMapper = $scopesToMapper;
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:38]: sort

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        $fullScopes = $scopeList;
+        foreach ($scopeList as $scope) {
+            \array_push($fullScopes, ...$this->getLowerHierarchyScopes($scope));
+        }
+        $fullScopes = \array_unique($fullScopes);
+
+        $acls = [];
+        foreach ($fullScopes as $scope) {
+            if (null === $scopeMapper = $this->getScopeMapper($scope)) {
+                continue;
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:103]: throw

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            return [];
+        }
+
+        return $scopeMapper->getLowerHierarchyScopes($scope);
+    }
+
+    private function getScopeMapper(string $scope): ?ScopeMapperInterface
+    {
+        if (!\array_key_exists($scope, $this->scopesToMapper)) {
+            return null;
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:130]: throw here

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface ScopeMapperInterface
+{
+    /**
+     * @return string[]
+     */
+    public function getAllScopes(): array;
```
Comment (willy-ahva) [inline src/Akeneo/Tool/Component/Api/Security/ScopeMapperInterface.php:16]:
```suggestion
    public function getScopes(): array;
```

File: src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php (hunk 6 above)
Comment (willy-ahva) [reply src/Akeneo/Tool/Component/Api/Security/ScopeMapperRegistry.php:38]: finally, it is not useful.

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    /** @var array<string, ScopeMapperInterface> */
+    private iterable $scopeMappers = [];
+
+    public function __construct(iterable $scopeMappers)
+    {
+        foreach ($scopeMappers as $scopeMapper) {
+            if (!$scopeMapper instanceof ScopeMapperInterface) {
+                throw new \InvalidArgumentException(
+                    \sprintf(
+                        '%s needs only %s',
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php:23]:
```suggestion
                        '%s should implement %s',
```

**Review by masacc — APPROVED** [review]

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                            get_class($this->scopeMappers[$scope])
+                        )
+                    );
+                }
+                $this->scopeMappers[$scope] = $scopeMapper;
+            }
+        }
+    }
+
+    public function getAllScopes(): array
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php:44]: @return string[]

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        return \array_keys($this->scopeMappers);
+    }
+
+    /**
+     * Provides all information needed to display scopes.
+     * Filters the scopes by keeping the higher hierarchy scope.
+     *
+     * @param string[] $scopeList
+     *
+     * @throw \LogicArgumentException if a scope you provide does not exist
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Apps/Security/ScopeMapperRegistry.php:55]:
```suggestion
     * @throw \LogicArgumentException if the given scope does not exist
```
so we won't confuse provide with providers

**Review by RiffVentura — APPROVED** [review]

**Pull request merged**